iroha/src/queue.rs

```diff
@@ -29,7 +46,11 @@ impl Queue {
     /// Makes queue from configuration
     pub fn from_configuration(cfg: &QueueConfiguration) -> Self {
         Self {
-            queue: ArrayQueue::new(cfg.maximum_transactions_in_queue as usize),
+            // See `get_transactions_for_block` implementation comment.
+            queue: ArrayQueue::new(
+                cfg.maximum_transactions_in_queue as usize
+                    + cfg.maximum_transactions_in_block as usize,
```

Comment:
Is it a bug though? Length of whole queue is measured by number of transactions in dashmap. So actually it should be always ok to push there. Did you encountered some error with that?

Reply:
There is a specific edge case - in the `get_transactions_for_block` we pop several transactions and sometimes push them back, assuming that queue has size to fit them. Concurrently with calling this function there might be queue pushes which will fill up the queue and our assumption will not be true for that case.

That's why public `push` will stop at `maximum_transactions_in_queue`. But queue will still have `maximum_transactions_in_block` places for the private push in `get_transactions_for_block`.

Reply:
This will never happen, as transactions are still will be in queue

Reply:
Yes, I see now, it won't happen as the push checks by number of transactions in the map which does not decrease there.